tortoise/fixture_test.go

```diff
@@ -0,0 +1,791 @@
+		val.before = append(val.before, a.reference)
+	}
+	if (val.Opinion.Hash == types.Hash32{}) {
+		val.Opinion.Hash = val.computeOpinion()
+	}
+
+	if val.base != nil {
+		val.before = append(val.before, val.base)
+	}
+	for _, atx := range val.activeset {
+		val.before = append(val.before, atx)
+	}
+	if len(val.activeset) == 0 {
+		val.before = append(val.before, a)
+	}
+
+	a.reg.register(val)
+	a.ballots[lid] = val
+	return val
+}
+
+type ballotAction struct {
+	types.BallotTortoiseData
+	base      *ballotAction
+	activeset []*atxAction
+	before    []action
+}
+
+func (b *ballotAction) String() string {
+	return fmt.Sprintf("ballot %v", b.ID)
+}
+
+func (b *ballotAction) deps() []action {
+	return b.before
+}
+
+func (b *ballotAction) computeOpinion() types.Hash32 {
+	votes := make([][]types.Vote, b.BallotTortoiseData.Layer.
+		Difference(types.GetEffectiveGenesis()),
+	) // vote on every layer before the ballot till effective genesis
+	for i := range votes {
+		votes[i] = []types.Vote{} // against by default
+	}
+	b.decodeVotes(votes)
+	hasher := opinionhash.New()
+	rst := types.Hash32{}
+	for i, layer := range votes {
+		if i != 0 {
+			hasher.WritePrevious(rst)
+		}
+		if layer == nil {
+			hasher.WriteAbstain()
+		}
+		if len(layer) > 0 {
+			sort.Slice(layer, func(i, j int) bool {
+				if layer[i].Height < layer[j].Height {
+					return true
+				}
+				return layer[i].ID.Compare(layer[j].ID)
+			})
+		}
+		for _, vote := range layer {
+			hasher.WriteSupport(vote.ID, vote.Height)
+		}
+		hasher.Sum(rst[:0])
+		hasher.Reset()
+	}
+	return rst
+}
+
+func (b *ballotAction) decodeVotes(votes [][]types.Vote) {
+	if b.base != nil {
+		b.base.decodeVotes(votes)
+	}
+	genesis := types.GetEffectiveGenesis()
+	for _, vote := range b.Opinion.Support {
+		pos := vote.LayerID.Difference(genesis)
+		matches := false
+		for _, existing := range votes[pos] {
+			if existing == vote {
+				matches = true
+			}
+		}
+		if !matches {
+			votes[pos] = append(votes[pos], vote)
+		}
+	}
+	for _, vote := range b.Opinion.Against {
+		pos := vote.LayerID.Difference(genesis)
+		for i, existing := range votes[pos] {
+			if existing == vote {
+				votes[pos] = append(votes[pos][:i], votes[pos][i:]...)
+				break
+			}
+		}
+	}
+	for _, vote := range b.Opinion.Abstain {
+		pos := vote.Difference(genesis)
+		votes[pos] = nil
+	}
+}
+
+func (b *ballotAction) execute(trt *Tortoise) {
+	decoded, err := trt.DecodeBallot(&b.BallotTortoiseData)
+	if err != nil {
+		panic(err)
+	}
+	if err := trt.StoreBallot(decoded); err != nil {
+		panic(err)
+	}
+}
+
+type action interface {
+	execute(*Tortoise)
+	deps() []action
+	String() string
+}
+
+func newSession(tb testing.TB) *session {
+	effective := types.GetEffectiveGenesis()
+	epochSize := types.GetLayersPerEpoch()
+	tb.Cleanup(func() {
+		types.SetEffectiveGenesis(effective.Uint32())
+		types.SetLayersPerEpoch(epochSize)
+	})
+	s := &session{tb: tb}
+	return s.
+		withLayerSize(50).
+		withEpochSize(10).
```

Comment:
`epochSize`?

Reply:
i didn't get it. can you clarify?

Reply:
i meant you already defined epochSize on line 316